pgx/src/datum/time_stamp_with_timezone.rs

```diff
@@ -8,143 +8,169 @@ Use of this source code is governed by the MIT license that can be found in the
+            }
+            TimestampWithTimeZone::INFINITY => Err(TimestampConversionError::TimestampInfinity),
+            _ => {
+                let sec = tstz.0 / USECS_PER_SEC;
+                let usec = tstz.0 - (sec * USECS_PER_SEC);
+                let duration = time::Duration::new(sec, (usec as i32) * 1000);
+                match PG_EPOCH_DATETIME.checked_add(duration) {
+                    Some(datetime) => Ok(datetime),
+                    None => Err(TimestampConversionError::TimeCrateConversionError),
+                }
+            }
+        }
     }
 }
 
-impl TimestampWithTimeZone {
-    /// This shifts the provided `time` back to UTC
-    pub fn new(time: time::PrimitiveDateTime, at_tz_offset: time::UtcOffset) -> Self {
-        TimestampWithTimeZone(
-            time.assume_utc()
-                .to_offset(
-                    UtcOffset::from_whole_seconds(-at_tz_offset.whole_seconds())
-                        .expect("Unexpected error in `UtcOffset::from_whole_seconds` during `TimestampWithTimeZone::new`")
-                ),
-        )
+impl TryFrom<time::PrimitiveDateTime> for TimestampWithTimeZone {
+    type Error = TimestampConversionError;
+
+    fn try_from(datetime: time::PrimitiveDateTime) -> Result<Self, Self::Error> {
+        let offset = datetime.assume_utc();
+        offset.try_into()
     }
 }
 
-impl Deref for TimestampWithTimeZone {
-    type Target = time::OffsetDateTime;
+impl TryFrom<TimestampWithTimeZone> for time::OffsetDateTime {
+    type Error = TimestampConversionError;
+    fn try_from(tstz: TimestampWithTimeZone) -> Result<Self, Self::Error> {
+        let datetime: time::PrimitiveDateTime = tstz.try_into()?;
+        Ok(datetime.assume_utc())
+    }
+}
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl IntoDatum for TimestampWithTimeZone {
+    fn into_datum(self) -> Option<pg_sys::Datum> {
+        Some(pg_sys::Datum::from(self.0))
+    }
+    fn type_oid() -> u32 {
+        pg_sys::TIMESTAMPTZOID
     }
 }
-impl DerefMut for TimestampWithTimeZone {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+
+impl FromDatum for TimestampWithTimeZone {
+    unsafe fn from_datum(datum: pg_sys::Datum, is_null: bool) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        if is_null {
+            None
+        } else {
+            Some(
+                datum
+                    .try_into()
+                    .expect("Error converting timestamp with time zone datum"),
+            )
+        }
     }
 }
 
+#[derive(thiserror::Error, Debug, Clone, Copy)]
+pub enum TimestampConversionError {
+    #[error("timestamp value is negative infinity and shouldn't map to time::PrimitiveDateTime")]
+    TimestampNegativeInfinity,
+    #[error("timestamp value is negative infinity and shouldn't map to time::PrimitiveDateTime")]
+    TimestampInfinity,
+    #[error("time::PrimitiveDateTime was unable to convert this timestamp")]
+    TimeCrateConversionError,
```

Comment:
This will almost always be written `TimestampConversionError::TimeCrateConversionError`, so we do not need to see a full spelling of `ConversionError` twice.